File: validator/auth/verify.cpp
#include <algorithm>
#include <set>

#include "verify.h"
namespace tos::auth {
namespace {
Result<std::uint64_t> checked_weight_add(std::uint64_t a, std::uint64_t b) {
  std::uint64_t value;
  if (__builtin_add_overflow(a, b, &value))
    return Error{"weight"};
  return value;
}
Result<std::uint64_t> checked_weight_mul(std::uint64_t a, std::uint64_t b) {
  std::uint64_t value;
  if (__builtin_mul_overflow(a, b, &value))
    return Error{"weight"};
  return value;
}
}  // namespace

Result<Keyref> key_reference(const Key& k) {
  auto id = object_id("key", k);
  if (!id.ok())
    return id.error();
  return Keyref{k.suite_, k.parameters_, k.epoch_, id.value()};
}
Result<Bytes> signing_statement(const Duty& duty, const Record& record) {
  std::vector<Keyref> refs;
  for (const auto& c : record.components_)
    refs.push_back({c.suite_, c.parameters_, c.epoch_, c.key_id_});
  return encode(Statement{duty, record.identity_, std::move(refs)});
}
Result<bool> validate_payload(const Duty& duty, const Bytes& payload) {
  Bytes preimage{duty.role_};
  preimage.insert(preimage.end(), payload.begin(), payload.end());
  auto hash = digest("payload", preimage);
  if (!hash.ok())
    return hash.error();
  if (hash.value() != duty.payload_hash_)
    return Error{"payload-hash"};
  if (duty.role_ == 5) {
    auto intent = decode<Update>(payload);
    if (!intent.ok())
      return intent.error();
    if (intent.value().operation_ < 1 || intent.value().operation_ > 7 || intent.value().nonce_ != duty.position_)
      return Error{"admin-payload"};
    return true;
  }
  if (duty.role_ < 1 || duty.role_ > 4 || duty.position_ > 0xffffffffULL)
    return Error{"role-position"};
  constexpr std::array<std::uint8_t, 4> candidate{0x3f, 0xcd, 0x91, 0xb6}, notarize{0xa8, 0x05, 0xf6, 0xcd},
      finalize{0x05, 0xe1, 0xa7, 0x40}, skip{0x26, 0x1f, 0x6b, 0x2f};
  auto role = duty.role_;
  std::size_t size = role == 1 ? 40 : role == 4 ? 8 : 44, offset = role == 2 || role == 3 ? 8 : 4;
  const auto& prefix = role == 1 ? candidate : role == 2 ? notarize : role == 3 ? finalize : skip;
  if (payload.size() != size || !std::equal(prefix.begin(), prefix.end(), payload.begin()))
    return Error{"payload-type"};
  if (offset == 8 && !std::equal(candidate.begin(), candidate.end(), payload.begin() + 4))
    return Error{"payload-type"};
  std::uint32_t slot = 0;
  for (unsigned i = 0; i < 4; ++i)
    slot |= std::uint32_t(payload[offset + i]) << (8 * i);
  if (slot != duty.position_)
    return Error{"payload-position"};
  return true;
}
Result<RegistrySnapshot> RegistrySnapshot::compile(const Committee& committee, const Policy& policy) {
  auto policy_raw = encode(policy);
  if (!policy_raw.ok())
    return policy_raw.error();
  if (policy.phase_ != 0 || policy.suites_ != std::vector<Suite>{{1, 1}})
    return Error{"unsupported-profile"};
  if (policy.interface_digest_ != interface_fingerprint)
    return Error{"interface-digest"};
  if (policy.revision_ == 0 || policy.max_envelope_ != 4096 || policy.max_certificate_ != 524288 ||
      policy.effective_from_ > committee.anchor_mc_)
    return Error{"c0-policy"};
  auto cid = object_id("committee", committee);
  if (!cid.ok())
    return cid.error();
  auto pid = object_id("policy", policy);
  if (!pid.ok())
    return pid.error();
  if (committee.policy_ != pid.value())
    return Error{"registry-policy"};
  if (committee.members_.empty() || committee.members_.size() > 400)
    return Error{"registry-bound"};
  RegistrySnapshot out;
  out.committee_ = committee;
  out.policy_ = policy;
  out.committee_id_ = cid.value();
  out.policy_id_ = pid.value();
  std::set<Hash> stakes;
  Hash previous{};
  for (const auto& member : committee.members_) {
    if (member.identity_ <= previous || member.stake_id_ == Hash{} || !stakes.insert(member.stake_id_).second)
      return Error{"registry-identities"};
    previous = member.identity_;
    auto total = checked_weight_add(out.total_, member.weight_);
    if (!total.ok())
      return total.error();
    if (member.weight_ == 0 || total.value() > max_weight)
      return Error{"weight"};
    out.total_ = total.value();
    if (member.keys_.size() != 5)
      return Error{"registry-keys"};
    Entry entry;
    entry.member = member;
    for (unsigned i = 0; i < 5; ++i) {
      const auto& k = member.keys_[i];
      if (k.role_ != i + 1 || k.suite_ != 1 || k.parameters_ != 1 || k.identity_ != member.identity_ || k.epoch_ == 0 ||
          k.valid_from_ > committee.anchor_mc_ || committee.anchor_mc_ >= k.valid_until_)
        return Error{"key-validity"};
      if (k.capacity_domain_ != Hash{} || k.capacity_limit_ != 0)
        return Error{"c0-capacity"};
      auto admitted = AdmittedKey::admit(k.suite_, k.parameters_, k.public_key_);
      if (!admitted.ok())
        return admitted.error();
      auto ref = key_reference(k);
      if (!ref.ok())
        return ref.error();
      entry.references.push_back(ref.value());
      entry.keys.push_back(admitted.value());
    }
    out.roster_.emplace(member.identity_, std::move(entry));
  }
  return out;
}
Result<std::uint64_t> RegistrySnapshot::verify_records(const Duty& duty, const Bytes& payload,
                                                       const std::vector<Record>& records, const Duty& expected,
                                                       bool quorum) const {
  if (duty != expected)
    return Error{"expected-context"};
  if (duty.committee_ != committee_id_ || duty.policy_ != policy_id_ || duty.workchain_ != committee_.workchain_ ||
      duty.shard_ != committee_.shard_ || duty.anchor_mc_ != committee_.anchor_mc_ ||
      duty.catchain_ != committee_.catchain_)
    return Error{"context-binding"};
  auto body = validate_payload(duty, payload);
  if (!body.ok())
    return body.error();
  if (records.empty() || records.size() > roster_.size())
    return Error{"signer-order"};
  struct Pending {
    const AdmittedKey* key;
    Bytes statement;
    const Bytes* signature;
  };
  std::vector<Pending> pending;
  Hash previous{};
  std::uint64_t weight = 0;
  for (const auto& row : records) {
    if (row.identity_ <= previous)
      return Error{"signer-order"};
    previous = row.identity_;
    auto member = roster_.find(row.identity_);
    if (member == roster_.end())
      return Error{"unknown-signer"};
    if (row.components_.size() != 1)
      return Error{"c0-components"};
    const auto& c = row.components_[0];
    if (Keyref{c.suite_, c.parameters_, c.epoch_, c.key_id_} != member->second.references[duty.role_ - 1])
      return Error{"key-binding"};
    if (c.signature_.size() != 64)
      return Error{"signature-size"};
    auto sum = checked_weight_add(weight, member->second.member.weight_);
    if (!sum.ok())
      return sum.error();
    if (sum.value() > max_weight)
      return Error{"weight"};
    weight = sum.value();
    auto statement = signing_statement(duty, row);
    if (!statement.ok())
      return statement.error();
    pending.push_back({&member->second.keys[duty.role_ - 1], std::move(statement.value()), &c.signature_});
  }
  auto signed_weight = checked_weight_mul(3, weight), required_weight = checked_weight_mul(2, total_);
  if (!signed_weight.ok())
    return signed_weight.error();
  if (!required_weight.ok())
    return required_weight.error();
  if (quorum && signed_weight.value() < required_weight.value())
    return Error{"quorum"};
  for (const auto& p : pending) {
    auto valid = p.key->verify(p.statement, *p.signature);
    if (!valid.ok())
      return valid.error();
    if (!valid.value())
      return Error{"signature"};
  }
  return weight;
}
Result<VerifiedCertificate> RegistrySnapshot::verify(const Certificate& cert, const Duty& expected) const {
  auto raw = encode(cert);
  if (!raw.ok())
    return raw.error();
  if (raw.value().size() > policy_.max_certificate_)
    return Error{"certificate-budget"};
  auto weight = verify_records(cert.duty_, cert.payload_, cert.records_, expected, true);
  if (!weight.ok())
    return weight.error();
  auto id = digest("certificate", raw.value());
  if (!id.ok())
    return id.error();
  std::vector<Hash> signers;
  for (const auto& r : cert.records_)
    signers.push_back(r.identity_);
  return VerifiedCertificate{id.value(), policy_id_, committee_id_, cert.duty_, std::move(signers), weight.value()};
}
Result<std::uint64_t> RegistrySnapshot::verify(const Envelope& env, const Duty& expected) const {
  auto raw = encode(env);
  if (!raw.ok())
    return raw.error();
  if (raw.value().size() > policy_.max_envelope_)
    return Error{"envelope-budget"};
  return verify_records(env.duty_, env.payload_, {env.record_}, expected, false);
}
}  // namespace tos::auth

File: tosctl/src/validator-auth/src/verify.rs
use crate::{
    codec::{decode, encode, Error, Hash},
    crypto::{digest, object_id, AdmittedKey},
    types::*,
};
use std::collections::{BTreeMap, BTreeSet};
pub const MAX_WEIGHT: u64 = u64::MAX / 3;
pub fn key_reference(k: &Key) -> Result<Keyref, Error> {
    Ok(Keyref {
        suite: k.suite,
        parameters: k.parameters,
        epoch: k.epoch,
        key_id: object_id("key", k)?,
    })
}
pub fn signing_statement(duty: &Duty, record: &Record) -> Result<Vec<u8>, Error> {
    encode(&Statement {
        duty: duty.clone(),
        identity: record.identity,
        keys: record
            .components
            .iter()
            .map(|c| Keyref {
                suite: c.suite,
                parameters: c.parameters,
                epoch: c.epoch,
                key_id: c.key_id,
            })
            .collect(),
    })
}
pub fn validate_payload(duty: &Duty, payload: &[u8]) -> Result<(), Error> {
    let mut bytes = vec![duty.role];
    bytes.extend_from_slice(payload);
    if digest("payload", &bytes)? != duty.payload_hash {
        return Err(Error("payload-hash"));
    }
    if duty.role == 5 {
        let intent = decode::<Update>(payload)?;
        if !(1..=7).contains(&intent.operation) || intent.nonce != duty.position {
            return Err(Error("admin-payload"));
        }
        return Ok(());
    }
    if !(1..=4).contains(&duty.role) || duty.position > u32::MAX as u64 {
        return Err(Error("role-position"));
    }
    let candidate = [0x3f, 0xcd, 0x91, 0xb6];
    let notarize = [0xa8, 0x05, 0xf6, 0xcd];
    let finalize = [0x05, 0xe1, 0xa7, 0x40];
    let skip = [0x26, 0x1f, 0x6b, 0x2f];
    let (size, prefix, offset) = match duty.role {
        1 => (40, candidate, 4),
        2 => (44, notarize, 8),
        3 => (44, finalize, 8),
        _ => (8, skip, 4),
    };
    if payload.len() != size || payload[..4] != prefix {
        return Err(Error("payload-type"));
    }
    if offset == 8 && payload[4..8] != candidate {
        return Err(Error("payload-type"));
    }
    let mut slot = [0; 4];
    slot.copy_from_slice(&payload[offset..offset + 4]);
    if u64::from(u32::from_le_bytes(slot)) != duty.position {
        return Err(Error("payload-position"));
    }
    Ok(())
}
#[derive(Clone, Debug)]
struct Entry {
    member: Member,
    references: Vec<Keyref>,
    keys: Vec<AdmittedKey>,
}
#[derive(Clone, Debug)]
pub struct RegistrySnapshot {
    committee: Committee,
    policy: Policy,
    committee_id: Hash,
    policy_id: Hash,
    roster: BTreeMap<Hash, Entry>,
    total: u64,
}
#[derive(Clone, Debug)]
pub struct VerifiedCertificate {
    certificate_id: Hash,
    duty: Duty,
    signers: Vec<Hash>,
    weight: u64,
}
impl VerifiedCertificate {
    pub fn certificate_id(&self) -> &Hash {
        &self.certificate_id
    }
    pub fn duty(&self) -> &Duty {
        &self.duty
    }
    pub fn signers(&self) -> &[Hash] {
        &self.signers
    }
    pub fn weight(&self) -> u64 {
        self.weight
    }
}
impl RegistrySnapshot {
    /// Admission validates the roster; native state proofs establish its authority.
    pub fn compile(committee: &Committee, policy: &Policy) -> Result<Self, Error> {
        if policy.phase != 0 || policy.suites != [Suite { suite: 1, parameters: 1 }] {
            return Err(Error("unsupported-profile"));
        }
        if policy.interface_digest != INTERFACE_FINGERPRINT {
            return Err(Error("interface-digest"));
        }
        if policy.revision == 0
            || policy.max_envelope != 4096
            || policy.max_certificate != 524288
            || policy.effective_from > committee.anchor_mc
        {
            return Err(Error("c0-policy"));
        }
        let policy_id = object_id("policy", policy)?;
        let committee_id = object_id("committee", committee)?;
        if committee.policy != policy_id {
            return Err(Error("registry-policy"));
        }
        if committee.members.is_empty() || committee.members.len() > 400 {
            return Err(Error("registry-bound"));
        }
        let mut out = Self {
            committee: committee.clone(),
            policy: policy.clone(),
            committee_id,
            policy_id,
            roster: BTreeMap::new(),
            total: 0,
        };
        let mut previous = [0; 32];
        let mut stakes = BTreeSet::new();
        for m in &committee.members {
            if m.identity <= previous || m.stake_id == [0; 32] || !stakes.insert(m.stake_id) {
                return Err(Error("registry-identities"));
            }
            previous = m.identity;
            let total = out.total.checked_add(m.weight).ok_or(Error("weight"))?;
            if m.weight == 0 || total > MAX_WEIGHT {
                return Err(Error("weight"));
            }
            out.total = total;
            if m.keys.len() != 5 {
                return Err(Error("registry-keys"));
            }
            let mut entry = Entry { member: m.clone(), references: Vec::new(), keys: Vec::new() };
            for (i, k) in m.keys.iter().enumerate() {
                if usize::from(k.role) != i + 1
                    || k.suite != 1
                    || k.parameters != 1
                    || k.identity != m.identity
                    || k.epoch == 0
                    || k.valid_from > committee.anchor_mc
                    || committee.anchor_mc >= k.valid_until
                {
                    return Err(Error("key-validity"));
                }
                if k.capacity_domain != [0; 32] || k.capacity_limit != 0 {
                    return Err(Error("c0-capacity"));
                }
                entry.keys.push(AdmittedKey::admit(&k.public_key)?);
                entry.references.push(key_reference(k)?);
            }
            out.roster.insert(m.identity, entry);
        }
        Ok(out)
    }
    pub fn committee_id(&self) -> &Hash {
        &self.committee_id
    }
    pub fn policy_id(&self) -> &Hash {
        &self.policy_id
    }
    pub fn committee(&self) -> &Committee {
        &self.committee
    }
    pub fn policy(&self) -> &Policy {
        &self.policy
    }
    pub fn total_weight(&self) -> u64 {
        self.total
    }
    fn verify_records(
        &self,
        duty: &Duty,
        payload: &[u8],
        rows: &[Record],
        expected: &Duty,
        quorum: bool,
    ) -> Result<u64, Error> {
        if duty != expected {
            return Err(Error("expected-context"));
        }
        if duty.committee != self.committee_id
            || duty.policy != self.policy_id
            || duty.workchain != self.committee.workchain
            || duty.shard != self.committee.shard
            || duty.anchor_mc != self.committee.anchor_mc
            || duty.catchain != self.committee.catchain
        {
            return Err(Error("context-binding"));
        }
        validate_payload(duty, payload)?;
        if rows.is_empty() || rows.len() > self.roster.len() {
            return Err(Error("signer-order"));
        }
        let mut weight = 0u64;
        let mut previous = [0; 32];
        let mut pending = Vec::new();
        for row in rows {
            if row.identity <= previous {
                return Err(Error("signer-order"));
            }
            previous = row.identity;
            let member = self.roster.get(&row.identity).ok_or(Error("unknown-signer"))?;
            if row.components.len() != 1 {
                return Err(Error("c0-components"));
            }
            let c = &row.components[0];
            let reference = Keyref {
                suite: c.suite,
                parameters: c.parameters,
                epoch: c.epoch,
                key_id: c.key_id,
            };
            if reference != member.references[usize::from(duty.role) - 1] {
                return Err(Error("key-binding"));
            }
            if c.signature.len() != 64 {
                return Err(Error("signature-size"));
            }
            let sum = weight.checked_add(member.member.weight).ok_or(Error("weight"))?;
            if sum > MAX_WEIGHT {
                return Err(Error("weight"));
            }
            weight = sum;
            pending.push((
                &member.keys[usize::from(duty.role) - 1],
                signing_statement(duty, row)?,
                &c.signature,
            ));
        }
        let signed_weight = weight.checked_mul(3).ok_or(Error("weight"))?;
        let required_weight = self.total.checked_mul(2).ok_or(Error("weight"))?;
        if quorum && signed_weight < required_weight {
            return Err(Error("quorum"));
        }
        for (key, statement, signature) in pending {
            if !key.verify(&statement, signature) {
                return Err(Error("signature"));
            }
        }
        Ok(weight)
    }
    pub fn verify_certificate(
        &self,
        cert: &Certificate,
        expected: &Duty,
    ) -> Result<VerifiedCertificate, Error> {
        let raw = encode(cert)?;
        if raw.len() > self.policy.max_certificate as usize {
            return Err(Error("certificate-budget"));
        }
        let weight =
            self.verify_records(&cert.duty, &cert.payload, &cert.records, expected, true)?;
        Ok(VerifiedCertificate {
            certificate_id: digest("certificate", &raw)?,
            duty: cert.duty.clone(),
            signers: cert.records.iter().map(|r| r.identity).collect(),
            weight,
        })
    }
    pub fn verify_envelope(&self, env: &Envelope, expected: &Duty) -> Result<u64, Error> {
        if encode(env)?.len() > self.policy.max_envelope as usize {
            return Err(Error("envelope-budget"));
        }
        self.verify_records(
            &env.duty,
            &env.payload,
            std::slice::from_ref(&env.record),
            expected,
            false,
        )
    }
}
